File: infra/src/main/java/com/beat/infra/persistence/schedule/repository/ScheduleRepositoryImpl.java
package com.beat.infra.persistence.schedule.repository;

import java.util.Collection;
import java.util.List;
import java.util.Optional;

import org.springframework.stereotype.Repository;

import com.beat.domain.schedule.domain.Schedule;
import com.beat.domain.schedule.repository.ScheduleRepository;
import com.beat.infra.persistence.schedule.entity.ScheduleJpaEntity;
import com.beat.infra.persistence.schedule.mapper.SchedulePersistenceMapper;

@Repository
public class ScheduleRepositoryImpl implements ScheduleRepository {

	private final ScheduleJpaRepository scheduleJpaRepository;
	private final SchedulePersistenceMapper schedulePersistenceMapper;

	public ScheduleRepositoryImpl(ScheduleJpaRepository scheduleJpaRepository,
		SchedulePersistenceMapper schedulePersistenceMapper) {
		this.scheduleJpaRepository = scheduleJpaRepository;
		this.schedulePersistenceMapper = schedulePersistenceMapper;
	}

	@Override
	public Optional<Schedule> findById(Long id) {
		return scheduleJpaRepository.findById(id)
			.map(schedulePersistenceMapper::toDomain);
	}

	@Override
	public Optional<Schedule> lockById(Long id) {
		return scheduleJpaRepository.lockById(id)
			.map(schedulePersistenceMapper::toDomain);
	}

	@Override
	public List<Schedule> findAllByPerformanceId(Long performanceId) {
		return scheduleJpaRepository.findAllByPerformanceId(performanceId).stream()
			.map(schedulePersistenceMapper::toDomain)
			.toList();
	}

	@Override
	public List<Schedule> findAllById(Collection<Long> ids) {
		return scheduleJpaRepository.findAllById(ids).stream()
			.map(schedulePersistenceMapper::toDomain)
			.toList();
	}

	@Override
	public List<Long> findIdsByPerformanceId(Long performanceId) {
		return scheduleJpaRepository.findIdsByPerformanceId(performanceId);
	}

	@Override
	public int countByPerformanceId(Long performanceId) {
		return scheduleJpaRepository.countByPerformanceId(performanceId);
	}

	@Override
	public Schedule save(Schedule schedule) {
		ScheduleJpaEntity savedEntity = scheduleJpaRepository.save(schedulePersistenceMapper.toEntity(schedule));
		return schedulePersistenceMapper.toDomain(savedEntity);
	}

	@Override
	public List<Schedule> saveAll(List<Schedule> schedules) {
		List<ScheduleJpaEntity> entities = schedules.stream()
			.map(schedulePersistenceMapper::toEntity)
			.toList();
		return scheduleJpaRepository.saveAll(entities).stream()
			.map(schedulePersistenceMapper::toDomain)
			.toList();
	}

	@Override
	public void delete(Schedule schedule) {
		if (schedule.getId() == null) {
			throw new IllegalArgumentException("Cannot delete unpersisted Schedule");
		}
		scheduleJpaRepository.deleteById(schedule.getId());
	}

	@Override
	public void deleteByPerformanceId(Long performanceId) {
		scheduleJpaRepository.deleteByPerformanceId(performanceId);
	}

	@Override
	public List<Schedule> findPendingSchedules() {
		return scheduleJpaRepository.findPendingSchedules().stream()
			.map(schedulePersistenceMapper::toDomain)
			.toList();
	}

}

File: core/infra/src/main/kotlin/com/beat/infra/persistence/schedule/repository/ScheduleRepositoryImpl.kt
package com.beat.infra.persistence.schedule.repository

import com.beat.domain.schedule.model.Schedule
import com.beat.domain.schedule.repository.ScheduleRepository
import com.beat.infra.persistence.schedule.mapper.SchedulePersistenceMapper
import org.springframework.stereotype.Repository
import org.springframework.data.repository.findByIdOrNull

@Repository
internal class ScheduleRepositoryImpl(
    private val scheduleJpaRepository: ScheduleJpaRepository,
    private val schedulePersistenceMapper: SchedulePersistenceMapper,
) : ScheduleRepository {

    override fun findById(id: Long): Schedule? =
        scheduleJpaRepository.findByIdOrNull(id)?.let(schedulePersistenceMapper::toDomain)

    override fun findPerformanceIdById(id: Long): Long? =
        scheduleJpaRepository.findPerformanceIdById(id)

    override fun lockById(id: Long): Schedule? =
        scheduleJpaRepository.lockById(id)?.let(schedulePersistenceMapper::toDomain)

    override fun isBeforeBookingCloseAt(id: Long): Boolean =
        scheduleJpaRepository.isBeforeBookingCloseAt(id) == 1L

    override fun findAllByPerformanceId(performanceId: Long): List<Schedule> =
        scheduleJpaRepository.findAllByPerformanceId(performanceId).map(schedulePersistenceMapper::toDomain)

    override fun findAllById(ids: Collection<Long>): List<Schedule> =
        scheduleJpaRepository.findAllById(ids).map(schedulePersistenceMapper::toDomain)

    override fun findIdsByPerformanceId(performanceId: Long): List<Long> =
        scheduleJpaRepository.findIdsByPerformanceId(performanceId)

    override fun countByPerformanceId(performanceId: Long): Int =
        scheduleJpaRepository.countByPerformanceId(performanceId)

    override fun save(schedule: Schedule): Schedule =
        scheduleJpaRepository.save(schedulePersistenceMapper.toEntity(schedule))
            .let(schedulePersistenceMapper::toDomain)

    override fun saveAll(schedules: List<Schedule>): List<Schedule> =
        scheduleJpaRepository.saveAll(schedules.map(schedulePersistenceMapper::toEntity))
            .map(schedulePersistenceMapper::toDomain)

    override fun delete(schedule: Schedule) {
        val scheduleId = requireNotNull(schedule.id) { "Cannot delete unpersisted Schedule" }
        scheduleJpaRepository.deleteById(scheduleId)
    }

    override fun deleteByPerformanceId(performanceId: Long) {
        scheduleJpaRepository.deleteByPerformanceId(performanceId)
    }

}
